### PR TITLE
(fix): refactor batch_processor to start on first process and use mutex to wait till deadline if nothing in the queue

### PR DESCRIPTION
**Pull request by thomaszurkan-optimizely**
## Summary
- Use a mutex so wait for a timeout since pop on SizedQueue does not support timeout.
- Empty the queue before going to mutex wait.
- Signal on process, flush, and stop.
- Add more debugging.

## Test plan
- Adjusted some unit tests.

### CODE REVIEW

File: spec/event/batch_event_processor_spec.rb [hunk 1]
```diff
@@ -62,7 +65,7 @@
 
     @event_processor.process(conversion_event)
     # flush interval is set to 100ms. Wait for 300ms and assert that event is dispatched.
-    sleep 1
+    sleep 0.2
```
Comment (oakbani) [inline spec/event/batch_event_processor_spec.rb]: nit. Comment on top no longer valid

File: lib/optimizely/event/batch_event_processor.rb [hunk 2]
```diff
@@ -147,15 +135,35 @@ def run
...
+        interval = (@flushing_interval_deadline - Helpers::DateTimeUtils.create_timestamp) * 0.001
+
+        next unless interval.positive?
+
+        @wait_mutex.synchronize { @resource.wait(@wait_mutex, interval) }
+      end
     rescue SignalException
       @logger.log(Logger::ERROR, 'Interrupted while processing buffer.')
-    rescue Exception => e
+    rescue => e
```
[the remark this answers is not in the document]
Comment (oakbani) [reply lib/optimizely/event/batch_event_processor.rb]: I just figured out that plain rescue => e means catching standard error. Looks good. 👍

File: .travis.yml [hunk 3]
```diff
@@ -9,7 +9,7 @@ rvm:
   - 2.5.1
   - 2.6.0
 before_install:
-  - gem update --system
+#  - gem update --system
```
Comment (mikeng13) [inline .travis.yml]: We don't need this anymore?

File: spec/event/batch_event_processor_spec.rb [hunk 4]
```diff
@@ -46,6 +47,8 @@
   it 'should log waring when service is already started' do
     @event_processor = Optimizely::BatchEventProcessor.new(logger: spy_logger)
     @event_processor.start!
+    @event_processor.start!
```
[the remark this answers is not in the document]
Comment (thomaszurkan-optimizely) [reply spec/event/batch_event_processor_spec.rb]: that is a test.  the test is in order to get the log message "already started"

**Pull request merged**